File: EclipseProjects/10FromPushToPull/test/pull/ValuesOfMonthTest.java
package pull;

import java.util.ArrayList;

import junit.framework.Assert;

import org.joda.time.LocalDate;
import org.junit.Test;

import common.Transaction;

public class ValuesOfMonthTest
{
	private static final LocalDate APRIL_ULTIMO = new LocalDate(2010, 4, 30);
	private static final LocalDate MAY_ULTIMO = new LocalDate(2010, 5, 31);

	@Test
	public void monthWithoutTransactionsHasZeroBalance()
	{
		ValuesOfMonth valuesOfMonth = new ValuesOfMonth(APRIL_ULTIMO, new ArrayList<Transaction>(), 0);
		Assert.assertEquals(0, valuesOfMonth.getBalance());
		Assert.assertEquals(0, valuesOfMonth.getAverageBalance());
	}

	@Test
	public void monthWithOneTransactionHasBalanceOfThat()
	{
		ArrayList<Transaction> transactionsOfMonth = new ArrayList<Transaction>();
		transactionsOfMonth.add(new Transaction(new LocalDate(2010, 4, 1), 100));

		ValuesOfMonth valuesOfMonth = new ValuesOfMonth(APRIL_ULTIMO, transactionsOfMonth, 0);
		Assert.assertEquals(100, valuesOfMonth.getBalance());
		Assert.assertEquals(100, valuesOfMonth.getAverageBalance());
	}

	@Test
	public void monthWithTwoTransactionsHasSumOfThoseAsBalance()
	{
		ArrayList<Transaction> transactionsOfMonth = new ArrayList<Transaction>();
		transactionsOfMonth.add(new Transaction(new LocalDate(2010, 4, 1), 100));
		transactionsOfMonth.add(new Transaction(new LocalDate(2010, 4, 16), 100));

		ValuesOfMonth valuesOfMonth = new ValuesOfMonth(APRIL_ULTIMO, transactionsOfMonth, 0);
		Assert.assertEquals(200, valuesOfMonth.getBalance());
		Assert.assertEquals(150, valuesOfMonth.getAverageBalance());
	}

	@Test
	public void twoMonthsWithTwoTransactionsHasSumOfThoseAsBalance()
	{
		ArrayList<Transaction> transactionsOfMonth = new ArrayList<Transaction>();
		transactionsOfMonth.add(new Transaction(new LocalDate(2010, 4, 16), 100));
		ValuesOfMonth april = new ValuesOfMonth(APRIL_ULTIMO, transactionsOfMonth, 0);

		transactionsOfMonth = new ArrayList<Transaction>();
		transactionsOfMonth.add(new Transaction(new LocalDate(2010, 5, 16), 200));
		ValuesOfMonth may = new ValuesOfMonth(MAY_ULTIMO, transactionsOfMonth, april.getBalance());

		Assert.assertEquals(300, may.getBalance());
		Assert.assertEquals(203, may.getAverageBalance());
	}
}

File: csharp/11_Pull.Test/ValuesOfMonthTest.cs
using System;
using System.Collections.Generic;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Common;

namespace Pull.Months.Test
{
    [TestClass]
    public class ValuesOfMonthTest
    {
        private static readonly DateTime APRIL_ULTIMO = new DateTime(2010, 4, 30);
        private static readonly DateTime MAY_ULTIMO = new DateTime(2010, 5, 31);

        [TestMethod]
        public void MonthWithoutTransactionsHasZeroBalance()
        {
            ValuesOfMonth valuesOfMonth = new ValuesOfMonth(APRIL_ULTIMO, new DummyValuesOfMonth());
            
            Assert.AreEqual(0, valuesOfMonth.Balance);
            Assert.AreEqual(0, valuesOfMonth.AverageBalance);
        }

        [TestMethod]
        public void MonthWithOneTransactionHasBalanceOfThat()
        {
            ValuesOfMonth valuesOfMonth = new ValuesOfMonth(APRIL_ULTIMO, new DummyValuesOfMonth());
            valuesOfMonth.AddTransaction(new Transaction { Date = new DateTime(2010, 4, 1), Amount = 100 });
            
            Assert.AreEqual(100, valuesOfMonth.Balance);
            Assert.AreEqual(100, valuesOfMonth.AverageBalance);
        }

        [TestMethod]
        public void MonthWithTwoTransactionsHasSumOfThoseAsBalance()
        {
            ValuesOfMonth valuesOfMonth = new ValuesOfMonth(APRIL_ULTIMO, new DummyValuesOfMonth());
            valuesOfMonth.AddTransaction(new Transaction { Date = new DateTime(2010, 4, 1), Amount = 100 });
            valuesOfMonth.AddTransaction(new Transaction { Date = new DateTime(2010, 4, 16), Amount = 100 });
            
            Assert.AreEqual(200, valuesOfMonth.Balance);
            Assert.AreEqual(150, valuesOfMonth.AverageBalance);
        }

        [TestMethod]
        public void TwoMonthsWithTwoTransactionsHasSumOfThoseAsBalance()
        {
            ValuesOfMonth april = new ValuesOfMonth(APRIL_ULTIMO, new DummyValuesOfMonth());
            april.AddTransaction(new Transaction { Date = new DateTime(2010, 4, 16), Amount = 100 });
            ValuesOfMonth may = new ValuesOfMonth(MAY_ULTIMO, april);
            may.AddTransaction(new Transaction { Date = new DateTime(2010, 5, 16), Amount = 200 });

            Assert.AreEqual(300, may.Balance);
            Assert.AreEqual(203, may.AverageBalance);
        }
    }
}
